### PR TITLE
session: add a test case for issue 19127

### LINKED ISSUE #19127: Affected rows are incorrect when updating a table with partitions
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table if exists t;
create table t (c_int int, c_str varchar(40), primary key (c_int, c_str) ) partition by hash (c_int) partitions 4;
insert into t values (9, 'angry williams'), (10, 'thirsty hugle');
update t set c_int = c_int + 10, c_str = 'adoring stonebraker' where c_int in (10, 9);
```

### 2. What did you expect to see? (Required)

Only 2 rows get affected.

### 3. What did you see instead (Required)

```
mysql> update t set c_int = c_int + 10, c_str = 'adoring stonebraker' where c_int in (10, 9);
Query OK, 4 rows affected (0.06 sec)
Rows matched: 2  Changed: 2  Warnings: 0
```
It seems the data is correct after update, only affected-rows in the response is wrong.

### 4. What is your TiDB version? (Required)

master @ 8978773f5e3d43a100550e6babea9904a99e5938
release-4.0 @ 7cac557b024f0aaca6825193477e5e6daea7055b

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #19127 

Problem Summary:

### What is changed and how it works?

What's Changed:

Add a test case to prevent regression.

How it Works:

The bug appears and disappears unexpectedly, so I want a test to cover the change.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

**Comment by tiancaiamao** [comment]
It's a bug;
**The behavior is correct by accident when the clustered index is enabled;**
When the clustered index feature is enable/disable during our development, the bug disappears or regression;
Without a test case to cover it, we did not recognize it.

The bug is caused by the mess operation of `stmtctx.AddAffectedRows`, sometimes we do it in [tables.AddRecord](https://github.com/pingcap/tidb/blob/master/table/tables/tables.go#L798) and sometimes we do it in [executor.updateRecord](https://github.com/pingcap/tidb/blob/5d8b99feef47bcae3eb51fff95c30ceabb55e2f5/executor/write.go#L232), when calling update, it might be repeat twice.

I believe handle `stmtctx.AddAffectedRows` in executor is better, because `UpdateRecord` may involve `RemoveRecord  + AddRecord` when the primary key change, in the table package's AddRecord it's lack of that context to know the following on RemoveRecord.

**Comment by tiancaiamao** [comment]
PTAL @cfzjywxk  @wshwsh12

**Comment by xhebox** [comment]
> I believe handle stmtctx.AddAffectedRows in executor is better

It makes sense. But do other SQLs other than `insert` depend on `AddAffectedRows` in `AddRecord`?

**Comment by tiancaiamao** [comment]
> > I believe handle stmtctx.AddAffectedRows in executor is better
> 
> It makes sense. But do other SQLs other than `insert` depend on `AddAffectedRows` in `AddRecord`?

No, this is the only place that  uses `AddAffectedRows()` in `table.AddRecord` @xhebox

File: executor/write_test.go [hunk 1]
```diff
@@ -1640,7 +1640,7 @@ func (s *testSuite4) TestPartitionedTableUpdate(c *C) {
 
 	// update partition column, old and new record locates on different partitions
 	tk.MustExec(`update t set id = 20 where id = 8`)
-	tk.CheckExecResult(2, 0)
+	tk.CheckExecResult(1, 0)
```
Comment (tiancaiamao) [inline executor/write_test.go:1643]: Here the old test case is wrong, this commit fix it.

**Comment by tangenta** [comment]
@tiancaiamao Should we cherry-pick to release-4.0 to prevent such failures in the AB test?

**Pull request merged**